### DISCUSSION TITLE
[Q] Pivot Table integration

### DISCUSSION THREAD

**Discussion by Shikate** [exceljs/exceljs#2432] [Q&A]
Hello, 

I'm currently using the lib to generate files through a dockerized API node server (with templating system, ...). 

I would like to be able to include tab with pivot table but, as mentioned in the documentation, it's not supported yet. 
It would be perfect if I could merge a tab from an .xlsx file (ie. my template), containing the pivot table, without breaking everything.
I don't need to modify this pivot table with exceljs, just duplication. 

Is there currently any way to do that ? 
If this is not the case I'll be glad to help by contributing on this project, and any tips on how to do that will be appreciated.

**Comment by fabrykowski** [comment]
See #261

**Comment by protobi-pieter** [comment]
Has anyone created a fork that implements pivot tables?   Or is anyone able/interested in doing so as a paid project?

 I too am interested in commissioning or adding support for pivot tables to **exceljs**.  
There are a few places in the xlsx OpenXML directory structure where pivot tables are specified.  
Below I've briefly noted the major differences between an xlsx file with a pivot table vs one without.

Defining and calculating the cache seems to be the major complication. It looks like exceljs would have to calculate the entire pivot table to write the results in both the worksheet and pivot table cache.  

My big question is whether it is necessary to calculate all the cached content or whether we can just save the definitions and Excel then calculates the pivot table on open (possibly by right clicking the pivot table in Excel GUI and clicking 'Refresh').  I experimented by selectively removing cached content and re-zipping the file but it appears that everything needs to be retained or excel Excel either complains on opening or the results that appear on opening don't reflect the source data.

- **workbook.xml** 
    - points to cached content via a relationshipId
   `<pivotCaches><pivotCache cacheId="5" r:id="rId3"/></pivotCaches>`
- **xl/worksheets/sheet1.xml**
   - Contains the calculated pivot table as regular worksheet cell content
- **xl/pivotTables/pivotTable/pivotTable1.xml**
   - Defines the pivot table
   ```
     <pivotTableDefinition>
         <pivotFields...>      
         <rowFields...>
         <rowItems...>
         <colItems...>
         <dataFields...>
         <pivotTableStyleInfo...>
         <extLst>
- **xl/pivotTables/pivotTable/_rels/pivotTable1.xml.rels**
   - points to the **pivotTableCache** definition
   `<Relationship Id="rId1" Type="http://schemas.openxmlformats.org/officeDocument/2006/relationships/pivotCacheDefinition" Target="../pivotCache/pivotCacheDefinition1.xml"/>`
- **xl/pivotCache/pivotCacheDefinition1.xml**
    - Defines the **pivotTableCache**
    ```
    <pivotCacheDefinition>
        <cacheSource><worksheetSource...>
        <cacheFields>
            <cacheField>
                  <sharedItems...> 
                         <n...>
        <extField>
    ```
- **xl/pivotCache/pivotCacheRecords1.xml**
     - Contains the actual cached numbers
     `<pivotCacheRecords><r><x><n>...`
- **xl/pivotCache/_rels/pivotCacheDefinition1.xml.rels**
    - points the pivotTableCacheDefinition to the pivotCacheRecords
    `<Relationship Id="rId1" Type="http://schemas.openxmlformats.org/officeDocument/2006/relationships/pivotCacheRecords" Target="pivotCacheRecords1.xml"/>`

**Comment by protobi-pieter** [comment]
See comment https://github.com/exceljs/exceljs/issues/2293#issuecomment-1697874719 for a way to add simple pivot tables.
